packages/mainmenu-extension/src/builder.ts

```diff
@@ -0,0 +1,287 @@
+import { Dialog, showDialog } from '@jupyterlab/apputils';
+import { JupyterLabMenu, IMainMenu } from '@jupyterlab/mainmenu';
+import { ISettingRegistry, SettingRegistry } from '@jupyterlab/settingregistry';
+
+import { ITranslator, TranslationBundle } from '@jupyterlab/translation';
+
+import { JSONExt } from '@lumino/coreutils';
+
+import { Menu } from '@lumino/widgets';
+
+export const PLUGIN_ID = '@jupyterlab/mainmenu-extension:plugin';
+
+export async function loadSettingsMenu(
+  registry: ISettingRegistry,
+  attachMenu: (menu: Menu) => void,
+  menuFactory: (options: IMainMenu.IMenuOptions) => JupyterLabMenu,
+  translator: ITranslator
+): Promise<void> {
+  const trans = translator.load('jupyterlab');
+  let canonical: ISettingRegistry.ISchema | null;
+  let loaded: { [name: string]: ISettingRegistry.IMenu[] } = {};
+
+  /**
+   * Populate the plugin's schema defaults.
+   */
+  function populate(schema: ISettingRegistry.ISchema) {
+    loaded = {};
+    schema.properties!.menus.default = Object.keys(registry.plugins)
+      .map(plugin => {
+        const menus =
+          registry.plugins[plugin]!.schema['jupyter.lab.menus']?.main ?? [];
+        loaded[plugin] = menus;
+        return menus;
+      })
+      .concat([schema.properties!.menus.default as any[]])
+      .reduceRight(
+        (acc, val) => SettingRegistry.reconcileMenus(acc, val, true),
+        []
+      ) // flatten one level
+      .sort((a, b) => (a.rank ?? Infinity) - (b.rank ?? Infinity));
+
+    schema.properties!.menus.description = trans.__(
+      `Note: To disable a menu or a menu item,
+copy it to User Preferences and add the
+"disabled" key. The following example will disable
+the "Tabs" menu and "Restart Kernel and Run up to Selected Cell"
+item:
+{
+  "menus": [
+    {
+      "id": "jp-mainmenu-tabs",
+      "disabled": true
+    },
+    {
+        "id": "jp-mainmenu-kernel",
+        "items": [
+            {
+                "command": "notebook:restart-and-run-to-selected",
+                "disabled": true
+            }
+        ]
```

Comment:
Use consistent indent here (either two or four spaces) to make it easier to understand.

Reply:
Thanks